akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -136,3 +139,58 @@ private[akka] final class SourceQueueAdapter[T](delegate: SourceQueue[T]) extend
   def offer(elem: T): CompletionStage[QueueOfferResult] = delegate.offer(elem).toJava
   def watchCompletion(): CompletionStage[Done] = delegate.watchCompletion().toJava
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class BlockingStreamSource[T, S](create: () ⇒ S,
+                                                       readData: (S) ⇒ T,
```

Comment:
> The read function needs to be able to signal stream completion, and since we don’t like null this will have to be an S => Option[T]

API will look better but example will be more verbose. Maybe will find out some elegant way to express in example